File: flightnotify/web/chart.py
"""Server-rendered price-history chart.

The SVG is generated on the server and scales with its container, so the chart
works with JavaScript disabled, resizes without clipping, and is always
accompanied by the same data as a real table (rendered by the template).
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from html import escape
from zoneinfo import ZoneInfo

from ..domain.pricing import format_money
from ..timeutil import to_local

VIEW_WIDTH = 720
VIEW_HEIGHT = 260
PADDING_LEFT = 62
PADDING_RIGHT = 16
PADDING_TOP = 18
PADDING_BOTTOM = 34


@dataclass(frozen=True, slots=True)
class PricePoint:
    observed_at: datetime
    amount: Decimal
    label: str


def render_price_chart(
    points: list[PricePoint],
    *,
    currency: str,
    tz: ZoneInfo,
    threshold: Decimal | None = None,
    low: Decimal | None = None,
    title: str = "Observed price history",
) -> str:
    """Return a self-contained, responsive, accessible SVG string."""
    if not points:
        return (
            '<p class="chart-empty">No successful observations yet, so there is nothing '
            "to chart.</p>"
        )

    ordered = sorted(points, key=lambda p: p.observed_at)
    values = [float(p.amount) for p in ordered]
    candidates = list(values)
    if threshold is not None:
        candidates.append(float(threshold))
    y_min, y_max = min(candidates), max(candidates)
    if y_max == y_min:
        y_max = y_min + max(1.0, y_min * 0.05)
    span = y_max - y_min
    y_min -= span * 0.08
    y_max += span * 0.08

    plot_w = VIEW_WIDTH - PADDING_LEFT - PADDING_RIGHT
    plot_h = VIEW_HEIGHT - PADDING_TOP - PADDING_BOTTOM

    def x_at(index: int) -> float:
        if len(ordered) == 1:
            return PADDING_LEFT + plot_w / 2
        return PADDING_LEFT + plot_w * index / (len(ordered) - 1)

    def y_at(value: float) -> float:
        ratio = (value - y_min) / (y_max - y_min)
        return PADDING_TOP + plot_h * (1 - ratio)

    parts: list[str] = []
    # Horizontal gridlines + y labels
    for step in range(4):
        value = y_min + (y_max - y_min) * step / 3
        y = y_at(value)
        parts.append(
            f'<line class="grid" x1="{PADDING_LEFT}" y1="{y:.1f}" '
            f'x2="{VIEW_WIDTH - PADDING_RIGHT}" y2="{y:.1f}" />'
        )
        parts.append(
            f'<text class="axis" x="{PADDING_LEFT - 8}" y="{y + 4:.1f}" text-anchor="end">'
            f"{escape(format_money(Decimal(str(round(value, 2))), currency))}</text>"
        )

    if threshold is not None:
        y = y_at(float(threshold))
        if PADDING_TOP <= y <= PADDING_TOP + plot_h:
            parts.append(
                f'<line class="threshold" x1="{PADDING_LEFT}" y1="{y:.1f}" '
                f'x2="{VIEW_WIDTH - PADDING_RIGHT}" y2="{y:.1f}" />'
            )
            parts.append(
                f'<text class="axis threshold-label" x="{VIEW_WIDTH - PADDING_RIGHT}" '
                f'y="{y - 6:.1f}" text-anchor="end">Threshold</text>'
            )

    coords = [(x_at(i), y_at(v)) for i, v in enumerate(values)]
    path = " ".join(
        ("M" if i == 0 else "L") + f"{x:.1f},{y:.1f}" for i, (x, y) in enumerate(coords)
    )
    area = (
        path
        + f" L{coords[-1][0]:.1f},{PADDING_TOP + plot_h:.1f}"
        + f" L{coords[0][0]:.1f},{PADDING_TOP + plot_h:.1f} Z"
    )
    parts.append(f'<path class="area" d="{area}" />')
    parts.append(f'<path class="line" d="{path}" />')

    low_value = float(low) if low is not None else min(values)
    for (x, y), point in zip(coords, ordered, strict=True):
        is_low = abs(float(point.amount) - low_value) < 0.005
        cls = "dot dot-low" if is_low else "dot"
        stamp = to_local(point.observed_at, tz)
        stamp_text = stamp.strftime("%b %-d, %-I:%M %p") if stamp else ""
        parts.append(
            f'<circle class="{cls}" cx="{x:.1f}" cy="{y:.1f}" r="{4.5 if is_low else 3}">'
            f"<title>{escape(format_money(point.amount, currency))} — "
            f"{escape(stamp_text)} — {escape(point.label)}</title></circle>"
        )

    # X labels: first, middle, last only, so they never collide.
    label_indexes = {0, len(ordered) - 1}
    if len(ordered) > 2:
        label_indexes.add(len(ordered) // 2)
    for index in sorted(label_indexes):
        stamp = to_local(ordered[index].observed_at, tz)
        if stamp is None:
            continue
        anchor = "start" if index == 0 else ("end" if index == len(ordered) - 1 else "middle")
        x = x_at(index)
        if anchor == "start":
            x = PADDING_LEFT
        elif anchor == "end":
            x = VIEW_WIDTH - PADDING_RIGHT
        parts.append(
            f'<text class="axis" x="{x:.1f}" y="{VIEW_HEIGHT - 10}" text-anchor="{anchor}">'
            f"{escape(stamp.strftime('%b %-d'))}</text>"
        )

    description = (
        f"Line chart of {len(ordered)} observed fares in {currency}, "
        f"from {escape(format_money(Decimal(str(min(values))), currency))} to "
        f"{escape(format_money(Decimal(str(max(values))), currency))}. "
        "The same data is listed in the table below."
    )
    return (
        f'<svg class="price-chart" viewBox="0 0 {VIEW_WIDTH} {VIEW_HEIGHT}" '
        'preserveAspectRatio="xMidYMid meet" role="img" '
        f'aria-label="{escape(title)}. {description}">'
        f"<desc>{description}</desc>" + "".join(parts) + "</svg>"
    )

File: worker/src/web/chart.ts
/**
 * Server-rendered price-history chart.
 *
 * Port of `flightnotify/web/chart.py`. The SVG is generated on the server and
 * scales with its container, so the chart works with JavaScript disabled,
 * resizes without clipping, and is always accompanied by the same data as a
 * real table (rendered by the view).
 *
 * Two deliberate departures from the Python source, both invisible in output:
 *
 * - Money arrives as integer cents (the Worker never stores floats), so
 *   geometry converts to dollar floats up front. Pixel positions never flow
 *   back into storage, and the degenerate-range fallback below is "one
 *   dollar", which only means the right thing in major units.
 * - `strftime` does not exist here; Intl.DateTimeFormat parts are reassembled
 *   into the same "%b %-d, %-I:%M %p" shapes, so English month abbreviations
 *   come from CLDR rather than a hand-rolled table.
 *
 * The return value is a raw markup string: every interpolated value is escaped
 * here (labels carry airline names from the provider), and the caller wraps
 * the whole thing in `raw()`.
 */

import { formatMoney } from "../domain/money.js";
import { parseIsoOrNull } from "../time.js";
import { escapeHtml } from "./html.js";

export interface PricePoint {
  /** ISO-8601 timestamp, UTC (the canonical stored shape). */
  observedAt: string;
  /** Integer minor units, as everywhere else in the Worker. */
  amountCents: number;
  /** Free text from the fare provider -- airline names, so never trusted. */
  label: string;
}

export interface PriceChartOptions {
  currency: string;
  timeZone: string;
  thresholdCents?: number | null;
  lowCents?: number | null;
  title?: string;
}

const VIEW_WIDTH = 720;
const VIEW_HEIGHT = 260;
const PADDING_LEFT = 62;
const PADDING_RIGHT = 16;
const PADDING_TOP = 18;
const PADDING_BOTTOM = 34;

/**
 * Ordering key. Canonical stored timestamps would sort as strings, but points
 * can be assembled from older rows; parsing keeps chronology authoritative.
 * An unparseable stamp sorts first rather than throwing -- the chart is
 * presentation, not validation.
 */
function timeValue(observedAt: string): number {
  return parseIsoOrNull(observedAt)?.getTime() ?? Number.NEGATIVE_INFINITY;
}

function partsFor(date: Date, timeZone: string, withTime: boolean): Map<string, string> {
  const parts = new Intl.DateTimeFormat("en-US", {
    timeZone,
    month: "short",
    day: "numeric",
    ...(withTime ? { hour: "numeric", minute: "2-digit", hour12: true } : {}),
  }).formatToParts(date);
  return new Map(parts.map((p) => [p.type, p.value]));
}

/**
 * Python's `%b %-d, %-I:%M %p` ("Aug 3, 2:15 PM"). Assembled from parts
 * because `format()` would insert ICU's narrow no-break space before AM/PM,
 * and because `formatLocal` in time.ts appends a timezone abbreviation the
 * original chart never showed.
 */
function formatStamp(date: Date, timeZone: string): string {
  const get = partsFor(date, timeZone, true);
  return (
    `${get.get("month")} ${get.get("day")}, ` +
    `${get.get("hour")}:${get.get("minute")} ${get.get("dayPeriod")}`
  );
}

/** Python's `%b %-d` ("Aug 3"), for the x axis. */
function formatDay(date: Date, timeZone: string): string {
  const get = partsFor(date, timeZone, false);
  return `${get.get("month")} ${get.get("day")}`;
}

/** Mirror of Python's `f"{value:.1f}"` for pixel coordinates. */
function px(value: number): string {
  return value.toFixed(1);
}

/** Return a self-contained, responsive, accessible SVG string. */
export function renderPriceChart(points: PricePoint[], options: PriceChartOptions): string {
  const {
    currency,
    timeZone,
    thresholdCents = null,
    lowCents = null,
    title = "Observed price history",
  } = options;

  if (points.length === 0) {
    return (
      '<p class="chart-empty">No successful observations yet, so there is nothing ' +
      "to chart.</p>"
    );
  }

  const ordered = [...points].sort((a, b) => timeValue(a.observedAt) - timeValue(b.observedAt));
  const values = ordered.map((p) => p.amountCents / 100);

  // The threshold joins the y-domain so a threshold far above every fare still
  // lands on the canvas instead of silently vanishing.
  const candidates = [...values];
  if (thresholdCents !== null) candidates.push(thresholdCents / 100);
  let yMin = Math.min(...candidates);
  let yMax = Math.max(...candidates);
  if (yMax === yMin) {
    // A flat series still needs vertical room, or every dot sits on one line
    // and the 8% padding below divides by zero span.
    yMax = yMin + Math.max(1.0, yMin * 0.05);
  }
  const span = yMax - yMin;
  yMin -= span * 0.08;
  yMax += span * 0.08;

  const plotW = VIEW_WIDTH - PADDING_LEFT - PADDING_RIGHT;
  const plotH = VIEW_HEIGHT - PADDING_TOP - PADDING_BOTTOM;

  const xAt = (index: number): number => {
    // A lone observation is centered rather than pinned to the left edge.
    if (ordered.length === 1) return PADDING_LEFT + plotW / 2;
    return PADDING_LEFT + (plotW * index) / (ordered.length - 1);
  };
  const yAt = (value: number): number => {
    const ratio = (value - yMin) / (yMax - yMin);
    return PADDING_TOP + plotH * (1 - ratio);
  };

  const parts: string[] = [];

  // Horizontal gridlines + y labels.
  for (let step = 0; step < 4; step += 1) {
    const value = yMin + ((yMax - yMin) * step) / 3;
    const y = yAt(value);
    parts.push(
      `<line class="grid" x1="${PADDING_LEFT}" y1="${px(y)}" ` +
        `x2="${VIEW_WIDTH - PADDING_RIGHT}" y2="${px(y)}" />`,
    );
    parts.push(
      `<text class="axis" x="${PADDING_LEFT - 8}" y="${px(y + 4)}" text-anchor="end">` +
        `${escapeHtml(formatMoney(Math.round(value * 100), currency))}</text>`,
    );
  }

  if (thresholdCents !== null) {
    const y = yAt(thresholdCents / 100);
    // Defensive: the domain above already contains the threshold, but a future
    // domain tweak must degrade to "no line", never to a line off-canvas.
    if (y >= PADDING_TOP && y <= PADDING_TOP + plotH) {
      parts.push(
        `<line class="threshold" x1="${PADDING_LEFT}" y1="${px(y)}" ` +
          `x2="${VIEW_WIDTH - PADDING_RIGHT}" y2="${px(y)}" />`,
      );
      parts.push(
        `<text class="axis threshold-label" x="${VIEW_WIDTH - PADDING_RIGHT}" ` +
          `y="${px(y - 6)}" text-anchor="end">Threshold</text>`,
      );
    }
  }

  const coords = values.map((v, i) => [xAt(i), yAt(v)] as const);
  const path = coords.map(([x, y], i) => `${i === 0 ? "M" : "L"}${px(x)},${px(y)}`).join(" ");
  const first = coords[0]!;
  const last = coords[coords.length - 1]!;
  const area =
    path +
    ` L${px(last[0])},${px(PADDING_TOP + plotH)}` +
    ` L${px(first[0])},${px(PADDING_TOP + plotH)} Z`;
  parts.push(`<path class="area" d="${area}" />`);
  parts.push(`<path class="line" d="${path}" />`);

  // The near-equality window matches Python's float comparison: the stored low
  // and the observation it came from can differ by float noise, never by half
  // a cent or more.
  const lowValue = lowCents !== null ? lowCents / 100 : Math.min(...values);
  for (let i = 0; i < ordered.length; i += 1) {
    const point = ordered[i]!;
    const [x, y] = coords[i]!;
    const isLow = Math.abs(values[i]! - lowValue) < 0.005;
    const cls = isLow ? "dot dot-low" : "dot";
    const stamp = parseIsoOrNull(point.observedAt);
    const stampText = stamp ? formatStamp(stamp, timeZone) : "";
    parts.push(
      `<circle class="${cls}" cx="${px(x)}" cy="${px(y)}" r="${isLow ? "4.5" : "3"}">` +
        `<title>${escapeHtml(formatMoney(point.amountCents, currency))} — ` +
        `${escapeHtml(stampText)} — ${escapeHtml(point.label)}</title></circle>`,
    );
  }

  // X labels: first, middle, last only, so they never collide.
  const labelIndexes = new Set<number>([0, ordered.length - 1]);
  if (ordered.length > 2) labelIndexes.add(Math.floor(ordered.length / 2));
  for (const index of [...labelIndexes].sort((a, b) => a - b)) {
    const stamp = parseIsoOrNull(ordered[index]!.observedAt);
    if (stamp === null) continue;
    const anchor = index === 0 ? "start" : index === ordered.length - 1 ? "end" : "middle";
    // Edge labels hug the frame instead of centering on their dot, so the
    // first and last never overflow the viewBox.
    let x = xAt(index);
    if (anchor === "start") x = PADDING_LEFT;
    else if (anchor === "end") x = VIEW_WIDTH - PADDING_RIGHT;
    parts.push(
      `<text class="axis" x="${px(x)}" y="${VIEW_HEIGHT - 10}" text-anchor="${anchor}">` +
        `${escapeHtml(formatDay(stamp, timeZone))}</text>`,
    );
  }

  const minCents = Math.min(...ordered.map((p) => p.amountCents));
  const maxCents = Math.max(...ordered.map((p) => p.amountCents));
  const description =
    `Line chart of ${ordered.length} observed fares in ${escapeHtml(currency)}, ` +
    `from ${escapeHtml(formatMoney(minCents, currency))} to ` +
    `${escapeHtml(formatMoney(maxCents, currency))}. ` +
    "The same data is listed in the table below.";
  return (
    `<svg class="price-chart" viewBox="0 0 ${VIEW_WIDTH} ${VIEW_HEIGHT}" ` +
    'preserveAspectRatio="xMidYMid meet" role="img" ' +
    `aria-label="${escapeHtml(title)}. ${description}">` +
    `<desc>${description}</desc>` +
    parts.join("") +
    "</svg>"
  );
}
